lib/URI.pm

```diff
@@ -123,7 +123,7 @@ sub _fix_uric_escape_for_host_part {
     return;
   }
 
-  if ($_[0] =~ m{^((?:$URI::scheme_re:)?)//([^/?\#]+)(.*)$}os) {
+  if ($_[0] =~ m{^((?:$URI::scheme_re:)?)//((.*:.*@)?[^/?\#]+)(.*)$}os) {
```

Comment:
Hm, but what if user wanna use '@' in password, using `[^@]+@` will cause that the everything before first '@' symbol will be matched and password parsed incorrectly.

like the similar approach is already used in [LWP/Protocol/http.pm](https://github.com/libwww-perl/libwww-perl/blob/1c1f28ddf472fbd9561b04be18601c89c9f833df/lib/LWP/Protocol/http.pm#L92).
And I already mentioned it in my [PR](https://github.com/libwww-perl/libwww-perl/pull/463#issuecomment-2287104200).

For my standpoint `(.*:.*@)` is pretty lazy, but okay, I agree that it could be better.

What you think about `([^:]+:.+@)` for matching user info?
Match everything except `:` as username cannot contain colon, then match `:` as separator, match everything (to be hones I didn't find where colon is forbidden for passwords), and finally match `@` as second separator.

Reply:
By the way, the regex which [Mojo/URL.pm#L52-64](https://metacpan.org/dist/Mojolicious/source/lib/Mojo/URL.pm#L52-64) is used also cannot parse username/password if it contains any of `/#?` symbols :) 
I know, it's really rarely usecase, but unfortunately I faced with it already two times :d

Reply:
Well, actually I can revert this change as the most important part is in [lib/URI/_server.pm#14](https://github.com/libwww-perl/URI/pull/144/files#diff-fbc962022819f5774ee8dd1a1e4d20c8ece9311d85af3a8b69d10546ac21cd87R14).

There are another issue, if password contain multiple '@' symbols URI parsing fails again :d

Reply:
> There are another issue, if password contain multiple '@' symbols before any of '/#?' symbol URI parsing fails again :d

If the password was inserted directly into the URL that way without percent-encoding, that's not really the URI module's fault.  The password has to be encoded properly first.  That's the point of escaping characters.